Randomized Smoothing Variations Implementation

f4str opened the pull request:
# Description

Implement the following variations of randomized smoothing

Fixes # (issue)

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

beat-buesser commented on art/estimators/certification/randomized_smoothing/smooth_mix/pytorch.py:
```diff
@@ -0,0 +1,337 @@
+        loss: "torch.nn.modules.loss._Loss",
+        input_shape: Tuple[int, ...],
+        nb_classes: int,
+        optimizer: Optional["torch.optim.Optimizer"] = None,  # type: ignore
```
```suggestion
        optimizer: Optional["torch.optim.Optimizer"] = None,
```

beat-buesser commented on art/estimators/certification/randomized_smoothing/smooth_mix/pytorch.py:
```diff
@@ -0,0 +1,337 @@
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        The authors' implementation of the SmoothMixPGD attack.
+        Code modified from https://github.com/jh-jeong/smoothmix/code/train.py
```
Let's include their MIT License text (https://github.com/jh-jeong/smoothmix/blob/main/LICENSE) at the top of our new file after ART's MIT license text.
```suggestion
        Code modified from https://github.com/jh-jeong/smoothmix/blob/main/code/train.py
```

beat-buesser reviewed (changes requested):
Hi @f4str Thank you very much for your pull request! The code looks great! I have a dded a few suggestions for minor improvements, what do you think?

f4str commented on art/estimators/certification/randomized_smoothing/macer/pytorch.py:
```diff
@@ -0,0 +1,238 @@
+        loss: "torch.nn.modules.loss._Loss",
+        input_shape: Tuple[int, ...],
+        nb_classes: int,
+        optimizer: Optional["torch.optim.Optimizer"] = None,  # type: ignore
```
The `type: ignore` is not required. The original `randomized_smoothing/pytorch.py`had this so I assumed it was needed. It can be removed from all variations including the original.

f4str commented:
Hi @beat-buesser thank you for the review. I've addressed your comments and made the corresponding changes. Feel free to let me know if there's any other issues.

beat-buesser reviewed (approved):
Hi @f4str Thank you very much for extending the randomised smoothing capabilities in ART! The changes look good to me.
